arroyo-connectors/src/filesystem.rs

```diff
@@ -0,0 +1,191 @@
+        grpc::api::Connector {
+            id: "filesystem".to_string(),
+            name: "FileSystem Sink".to_string(),
+            icon: ICON.to_string(),
+            description: "Write to a filesystem (S3)".to_string(),
+            enabled: true,
+            source: false,
+            sink: true,
+            testing: false,
+            custom_schemas: true,
+            connection_config: None,
+            table_config: TABLE_SCHEMA.to_owned(),
+        }
+    }
+
+    fn test(
+        &self,
+        _: &str,
+        _: Self::ConfigT,
+        _: Self::TableT,
+        _: Option<&arroyo_rpc::grpc::api::ConnectionSchema>,
+        tx: tokio::sync::mpsc::Sender<
+            Result<arroyo_rpc::grpc::api::TestSourceMessage, tonic::Status>,
+        >,
+    ) {
+        tokio::task::spawn(async move {
+            tx.send(Ok(TestSourceMessage {
+                error: false,
+                done: true,
+                message: "Successfully validated connection".to_string(),
+            }))
+            .await
+            .unwrap();
+        });
+    }
+
+    fn table_type(&self, _: Self::ConfigT, _: Self::TableT) -> grpc::api::TableType {
+        return grpc::api::TableType::Source;
+    }
+
+    fn from_config(
+        &self,
+        id: Option<i64>,
+        name: &str,
+        config: Self::ConfigT,
+        table: Self::TableT,
+        schema: Option<&ConnectionSchema>,
+    ) -> anyhow::Result<crate::Connection> {
+        let (description, operator) = match &table.format_settings {
+            Some(FormatSettings::Parquet { .. }) => ("FileSystem<Parquet>".to_string(), "connectors::filesystem::ParquetFileSystemSink::<#in_k, #in_t, #in_tRecordBatchBuilder>"),
+            Some(FormatSettings::Json {  }) => ("FileSystem<JSON>".to_string(), "connectors::filesystem::JsonFileSystemSink::<#in_k, #in_t>"),
+            None => bail!("have to have some format settings"),
+        };
+
+        let config = OperatorConfig {
+            connection: serde_json::to_value(config).unwrap(),
+            table: serde_json::to_value(table).unwrap(),
+            rate_limit: None,
+            serialization_mode: Some(serialization_mode(schema.as_ref().unwrap())),
+        };
+
+        Ok(Connection {
+            id,
+            name: name.to_string(),
+            connection_type: ConnectionType::Sink,
+            schema: schema
+                .map(|s| s.to_owned())
+                .ok_or_else(|| anyhow!("No schema defined for FileSystemSink"))?,
+            operator: operator.to_string(),
+            config: serde_json::to_string(&config).unwrap(),
+            description,
+        })
+    }
+
+    fn from_options(
+        &self,
+        name: &str,
+        opts: &mut std::collections::HashMap<String, String>,
+        schema: Option<&ConnectionSchema>,
+    ) -> anyhow::Result<crate::Connection> {
+        let write_target = if let (Some(s3_bucket), Some(s3_directory), Some(aws_region)) = (
```

Comment:
The error handling here isn't very user friendly, since if the user misses one of these options or mistypes them they get an unhelpful "Only s3 target are supported". See how error handling is done in the other connectors.

Reply:
Improved it, I think.